### DISCUSSION TITLE
pyvisa pulsed measurement AWG70k

### DISCUSSION THREAD

**Discussion by Maabur** [Ulm-IQO/qudi-core#49] [Q&A]
I would like to run the pulsed tool chain in the new core with the awg70k.
When I load the gui I have the following error message.
```
Qudi IPython kernel connected to local module service.
No ConfigOption "assets_storage_path" configured, using default value "C:\Users\IQO\saved_pulsed_assets" instead.
No ConfigOption "tmp_work_dir" configured, using default value "C:\Users\IQO\AppData\Roaming\qudi\pulsed_files" instead.
VISA address "XXXXXXX" not found by the pyVISA resource manager.
Check the connection by using for example "Agilent Connection Expert".
Exception during activation:
Traceback (most recent call last):
  File "C:\qudi core 2021 only\qudi 2021 core only dev\src\qudi\core\module.py", line 301, in __activation_callback
    self.on_activate()
  File "C:\Qudi 2021 core\qudi-iqo-modules\src\qudi\hardware\awg\tektronix_awg70k.py", line 120, in on_activate
    self.__max_seq_steps = int(self.query('SLIS:SEQ:STEP:MAX?'))
  File "C:\Qudi 2021 core\qudi-iqo-modules\src\qudi\hardware\awg\tektronix_awg70k.py", line 1250, in query
    return self.awg.query(question).strip().rstrip('\n').rstrip().strip('"')
AttributeError: 'NoneType' object has no attribute 'query'

Traceback (most recent call last):
  File "C:\qudi core 2021 only\qudi 2021 core only dev\src\qudi\core\modulemanager.py", line 201, in activate_module
    self._modules[module_name].activate()
  File "C:\qudi core 2021 only\qudi 2021 core only dev\src\qudi\core\modulemanager.py", line 502, in activate
    module.activate()
  File "C:\qudi core 2021 only\qudi 2021 core only dev\src\qudi\core\modulemanager.py", line 502, in activate
    module.activate()
  File "C:\qudi core 2021 only\qudi 2021 core only dev\src\qudi\core\modulemanager.py", line 502, in activate
    module.activate()
  File "C:\qudi core 2021 only\qudi 2021 core only dev\src\qudi\core\modulemanager.py", line 543, in activate
    raise RuntimeError(f'Failed to activate {self.module_base} module "{self.name}"!')
RuntimeError: Failed to activate hardware module "awg_tek"!
```

The old qudi was installed in a way that the test of the test of the pyvisa address is commented out in the hardware file. 
see here an extract of the activate function in the tektronix_awg70k.py hardware file.

```
 def on_activate(self):
        """ Initialisation performed during activation of the module.
        """
        # Create work directory if necessary
        if not os.path.exists(self._tmp_work_dir):
            os.makedirs(os.path.abspath(self._tmp_work_dir))

        # connect to awg using PyVISA
        #if self._visa_address not in self._rm.list_resources():
       #     self.awg = None
        #    self.log.error('VISA address "{0}" not found by the pyVISA resource manager.\nCheck '
      #                     'the connection by using for example "Agilent Connection Expert".'
     #                      ''.format(self._visa_address))
    #    else:
        self.awg = self._rm.open_resource(self._visa_address)
        # set timeout by default to 30 sec
        self.awg.timeout = self._visa_timeout * 1000
```

I think the address is correct. I run Qudi on the old core with this address as it is the one mentioned on the config file
here in the error message, I changed the visa address into XXXXX  for confidentiality reasons.

I tried to do the same on the new core but I get another error message.

**Comment by Neverhorst** [comment]
Well, from the traceback it is clear that you configured the `qudi.hardware.awg.tektronix_awg70k` hardware module ("awg_tek") as pulse generator for the pulsed measurement toolchain with an invalid VISA address string.
The IP address of your Tektronix AWG70k is probably wrong.

If you intend to start the pulsed measurement toolchain with a dummy pulse generator, you need to configure the pulser dummy module instead of the "real" hardware module (see `default.cfg` in the `qudi-iqo-modules` repository)

**Reply by Maabur** [reply]
Thanks, would you be able to check my original post ? I updated it was lacking some information.

**Reply by Neverhorst** [reply]
No code changes are needed. It works on other setups, so it's not a problem with the hardware module.

The new core has changed nothing in the AWG70k hardware module regarding VISA connections.
I'm sure you have the wrong config file loaded, or wrong VISA address, or something else botched with setting up the AWG connection.

Did you check the VISA address and functioning connection using e.g. "Keysight Connection Expert" or other VISA device manager software?

**Reply by Maabur** [reply]
I see, 
so Yeah I checked with the keysight connection expert and the visa address is the same as the one reported in the error message and the connection expert say that the device is connected (the green tick mark on the top of the device)

**Reply by timoML** [reply]
Can you enter to the manager and post the result?
```
import visa
visa.ResourceManager().list_resources()
```

**Reply by Maabur** [reply]
oh yeah right, it fails to import visa

**Reply by Neverhorst** [reply]
Then exchange `visa` with `pyvisa`

**Reply by Maabur** [reply]
when I import pyvisa as visa I get 3 different devices

**Reply by Neverhorst** [reply]
Well if you see the AWG address in there, please copy it exactly as reported by this command to the config file you are using.

**Reply by Maabur** [reply]
---

**Reply by Neverhorst** [reply]
Can you please try the same thing in "old qudi" where you said it works?

**Reply by Maabur** [reply]
in the old qudi, the command gives the same output.
I will do some more research and keep you posted

**Reply by Neverhorst** [reply]
Could it be that you are already connected to the device from some other software or qudi instance. In that case please close all other connections before trying to start qudi.

**Reply by Maabur** [reply]
So on the old qudi I can actually see the awg tcp address when I run the command visa.ResourceManager().list_resources() 
after loading the pulsed measurement gui. 
I managed to load the pulsed gui on the new core by commenting out the section like it is done on the hardware file in the old core and removing the tabs after the else:
        # connect to awg using PyVISA
        # if self._visa_address not in self._rm.list_resources():
        #     self.awg = None
        #     self.log.error('VISA address "{0}" not found by the pyVISA resource manager.\nCheck '
        #                    'the connection by using for example "Agilent Connection Expert".'
        #                    ''.format(self._visa_address))
        # else:
        self.awg = self._rm.open_resource(self._visa_address)
        # set timeout by default to 30 sec
        self.awg.timeout = self._visa_timeout * 1000

I can load the sequence and play a sequence bur when I run visa.ResourceManager().list_resources(), I do not get the awg tcp address

**Comment by liuyanulm** [comment]
None of these addresses look like the awg visa address. The bio-setup used
to have a keysight awg.
Better ask someone who is familiar with qudi come to the lab and fix it .
Maabur ***@***.***>于2023年5月10日 周三23:10写道：

**Reply by Maabur** [reply]
ok thanks
